persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,536 @@
+    , tokenContent
+    , anyToken
+    , ParsedEntityDef
+    , parsedEntityDefComments
+    , parsedEntityDefEntityName
+    , parsedEntityDefIsSum
+    , parsedEntityDefEntityAttributes
+    , parsedEntityDefFieldAttributes
+    , parsedEntityDefFieldComments
+    , parsedEntityDefExtras
+    , parsedEntityDefSpan
+    , parseSource
+    , memberBlockAttrs
+    ) where
+
+import Prelude hiding (lines, unlines)
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NEL
+import qualified Data.Map as M
+import Data.Maybe
+import Data.Text (Text, pack, unlines, unpack)
+import Data.Void
+import Database.Persist.Types
+import Database.Persist.Types.Span
+import Language.Haskell.TH.Syntax (Lift)
+import Replace.Megaparsec (sepCap)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+type Parser = Parsec Void String
+
+-- | Source location: file and line/col information. This is half of a 'Span'.
+data SourceLoc = SourceLoc
+    { locFile :: Text
+    , locStartLine :: Int
+    , locStartCol :: Int
+    }
+    deriving (Show, Lift)
+
+data Token
+    = DocComment Text
+    | Comment Text
+    | Quotation Text
+    | Equality Text Text
+    | Parenthetical Text
+    | BlockKey Text
+    | PText Text
+    deriving (Eq, Ord, Show)
+
+tokenContent :: Token -> Text
+tokenContent = \case
+    Comment s -> s
+    DocComment s -> s
+    Quotation s -> s
+    Equality l r -> mconcat [l, "=", r]
+    Parenthetical s -> s
+    PText s -> s
+    BlockKey s -> s
+
+spaceConsumer :: Parser ()
+spaceConsumer =
+    L.space
+        hspace1
+        (L.skipLineComment "--")
+        empty
+
+spaceConsumerN :: Parser ()
+spaceConsumerN =
+    L.space
+        space1
+        (L.skipLineComment "--")
+        empty
+
+escapedParen :: Parser Char
+escapedParen =
+    char '\\'
+        *> choice
+            [ char '('
+            , char ')'
+            ]
+
+contentChar :: Parser Char
+contentChar =
+    choice
+        [ alphaNumChar
+        , char '.'
+        , char '['
+        , char ']'
+        , char '_'
+        , char '\''
+        , char '"'
+        , char '!'
+        , char '~'
+        , char '-'
+        , char ':'
+        , try escapedParen
+        , char '\\'
+        ]
+
+nonLineSpaceChar :: Parser Char
+nonLineSpaceChar = choice [char ' ', char '\t']
+
+-- This is a replacement for `Text.Megaparsec.Char.Lexer.charLiteral`;
+-- it does nearly the same thing but additionally supports escaped parentheses.
+charLiteral :: Parser Char
+charLiteral = label "literal character" $ do
+    r <- lookAhead (count' 1 10 anySingle)
+    case lexChar r of
+        Just (c, r') -> c <$ skipCount (length r - length r') anySingle
+        Nothing -> unexpected (errorTokens r)
+  where
+    errorTokens ts = Tokens $
+        case NEL.nonEmpty ts of
+            Just nel -> NEL.head nel :| []
+            Nothing -> error "unreachable" -- ts is known to be non-empty
```

Comment:
I think we can do better here -

`NEL.nonEmpty :: [a] -> Maybe (NonEmpty a)`, right? So in `Just nel`, we already have a `nel :: NonEmpty a`. Doing `NEL.head nel :| []` is going to strip the first element off and create just a singleton `NonEmpty`, discarding every other element in the input list. That *feels* like maybe not what we want? At the very least, it suggests that our input may not be well formed here.

Let's imagine that we've made a mistake in our programming, and the `Nothing` case fires. We get `unreachable` as the only thing in our output to diagnose. Instead, let's encode the actual expectation in the message.

```suggestion
        case NEL.nonEmpty ts of
            Just nel -> NEL.head nel :| []
            Nothing -> error "errorTokens was somehow empty"
```

Reply:
> Doing NEL.head nel :| [] is going to strip the first element off and create just a singleton NonEmpty, discarding every other element in the input list. That feels like maybe not what we want? 

So this function is a modification of `charLiteral` in Megaparsec, and it's doing this too. This _is_ what we want, I guess, because `unexpected` expects a NEL and only the first element of that list is meaningful in this context. The underlying question here is why it's looking ahead ten characters.

Reply:
(My guess is that it's because whoever wrote that code also copy/pasted it from someplace else that was handling longer escape sequences)